new exception handler adoption

bkayes opened the pull request:
#### 9bdabedf3aab2b9b2b562d6f1978aacec147f747
<pre>
new exception handler adoption
<a href="https://bugs.webkit.org/show_bug.cgi?id=269728">https://bugs.webkit.org/show_bug.cgi?id=269728</a>
<a href="https://rdar.apple.com/119256062">rdar://119256062</a>

Reviewed by Keith Miller.

This adopts the new hardened mach exception mechanism which was
introduced in darwin recently. See the radar for more details.

* Source/JavaScriptCore/llint/LLIntData.cpp:
* Source/JavaScriptCore/llint/LLIntData.h:
* Source/JavaScriptCore/runtime/MachineContext.h:
(JSC::MachineContext::setInstructionPointer):
* Source/JavaScriptCore/wasm/WasmFaultSignalHandler.cpp:
(JSC::Wasm::trapHandler):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/PlatformJSCOnly.cmake:
* Source/WTF/wtf/PlatformMac.cmake:
* Source/WTF/wtf/PlatformRegisters.h:
* Source/WTF/wtf/threads/Signals.cpp:
(WTF::initMachExceptionHandlerThread):
(WTF::setExceptionPorts):
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:

Canonical link: <a href="https://commits.webkit.org/276198@main">https://commits.webkit.org/276198@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2322e94ab4579b601e0f5c9ac3bff48ea925f6c0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43959 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23026 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46397 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46610 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40034 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46264 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26930 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20420 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44537 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20057 "Found 2 new test failures: fast/forms/datalist/datalist-option-labels.html, fast/forms/datalist/datalist-textinput-dynamically-add-options-on-keydown.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37871 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17286 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17561 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38958 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2020 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/37335 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40142 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39250 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48184 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/43554 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18966 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15542 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43115 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20353 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41830 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9786 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20560 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/50600 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19979 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10211 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/runtime/InitializeThreading.cpp:
```diff
@@ -137,7 +137,15 @@ void initialize()
             // JSLock::lock() can call registerThreadForMachExceptionHandling() which crashes if this has not been called first.
-            initializeSignalHandling();
+            int mask = 0;
+#if CPU(ARM64E) && defined(EXCEPTION_STATE_IDENTITY_PROTECTED) && !PLATFORM(IOS_FAMILY_SIMULATOR)
```
Should this be `HAVE(HARDENED_MACH_EXCEPTIONS)`?

kmiller68 commented on Source/JavaScriptCore/wasm/WasmFaultSignalHandler.cpp:
```diff
@@ -38,12 +38,30 @@
+
+void* presignedTrampoline = { };
+
+#if CPU(ARM64E) && defined(EXCEPTION_STATE_IDENTITY_PROTECTED) && !PLATFORM(IOS_FAMILY_SIMULATOR)
```
Should this be `HAVE(HARDENED_MACH_EXCEPTIONS)`?

kmiller68 commented on Source/JavaScriptCore/wasm/WasmFaultSignalHandler.cpp:
```diff
@@ -38,12 +38,30 @@
 
+using WTF::CodePtr;
+
+void* presignedTrampoline = { };
```
I guess this could be pushed inside the `#if` below

kmiller68 commented on Source/JavaScriptCore/wasm/WasmFaultSignalHandler.cpp:
```diff
@@ -96,7 +114,11 @@ static SignalAction trapHandler(Signal signal, SigInfo& sigInfo, PlatformRegiste
             };
 
             if (didFaultInWasm(faultingInstruction)) {
+#if defined(EXCEPTION_STATE_IDENTITY_PROTECTED) && !PLATFORM(IOS_FAMILY_SIMULATOR)
```
Should this be `HAVE(HARDENED_MACH_EXCEPTIONS)` or at least include `CPU(ARM64E)`?

kmiller68 commented on Source/JavaScriptCore/wasm/WasmFaultSignalHandler.h:
```diff
@@ -37,4 +36,12 @@ inline void activateSignalingMemory() { }
 inline void prepareSignalingMemory() { }
 #endif // ENABLE(WEBASSEMBLY)
 
+#if CPU(ARM64E) && defined(EXCEPTION_STATE_IDENTITY_PROTECTED) && !PLATFORM(IOS_FAMILY_SIMULATOR)
```
Ditto.

kmiller68 commented on Source/WTF/wtf/PlatformRegisters.cpp:
```diff
@@ -53,6 +53,14 @@ void* threadStateLRInternal(PlatformRegisters& regs)
 
 void* threadStatePCInternal(PlatformRegisters& regs)
 {
+#if CPU(ARM64E) && defined(EXCEPTION_STATE_IDENTITY_PROTECTED) && !PLATFORM(IOS_FAMILY_SIMULATOR)
```
Ditto?

kmiller68 commented on Source/WTF/wtf/PlatformRegisters.h:
```diff
@@ -130,8 +131,13 @@ using WTF::threadStatePCInternal;
 #define WTF_READ_PLATFORM_REGISTERS_PC_WITH_PROFILE(regs) \
     threadStatePCInternal(const_cast<PlatformRegisters&>(regs))
 
+#if defined(EXCEPTION_STATE_IDENTITY_PROTECTED) && !PLATFORM(IOS_FAMILY_SIMULATOR)
```
Ditto?

pvollan commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -1817,9 +1817,14 @@
     thread_info
+#if HAVE(HARDENED_MACH_EXCEPTIONS)
+    task_register_hardened_exception_handler
+    thread_adopt_exception_handler
```
I think we need to check if these are defined:

    (when (defined? 'task_register_hardened_exception_handler) task_register_hardened_exception_handler)
    (when (defined? 'thread_adopt_exception_handler) thread_adopt_exception_handler)

pvollan commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -1839,7 +1844,11 @@
 #if ENABLE(BLOCK_SET_EXCEPTION_PORTS)
             (with-filter (require-not (webcontent-process-launched))
+#if HAVE(HARDENED_MACH_EXCEPTIONS)
+                (allow mach-message-send (kernel-mig-routine task_register_hardened_exception_handler)))
```
Ditto.

            (when (defined? 'task_register_hardened_exception_handler)
                (allow mach-message-send (kernel-mig-routine task_register_hardened_exception_handler))))

bkayes replied:
can I do `defined?` and a `require-not`?

pvollan commented on Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
```diff
@@ -1466,13 +1466,20 @@
 #endif
 
+#if HAVE(HARDENED_MACH_EXCEPTIONS)
+        (when (defined? 'task_register_hardened_exception_handler) (with-filter (require-not (webcontent-process-launched))
```
This check is not needed on iOS, I believe. I think the previous version was correct:

#if HAVE(HARDENED_MACH_EXCEPTIONS)
#if ENABLE(BLOCK_SET_EXCEPTION_PORTS)
        (with-filter (require-not (webcontent-process-launched))
            (allow mach-message-send (kernel-mig-routine task_register_hardened_exception_handler)))
#endif ;; ENABLE(BLOCK_SET_EXCEPTION_PORTS)
        (with-filter (require-not (lockdown-mode))
            (allow mach-message-send (kernel-mig-routine thread_adopt_exception_handler)))
#else
#if ENABLE(BLOCK_SET_EXCEPTION_PORTS)
        (with-filter (require-not (webcontent-process-launched))
            (allow mach-message-send (kernel-mig-routine thread_set_exception_ports)))
#else
        (with-filter (require-not (lockdown-mode))
            (allow mach-message-send (kernel-mig-routine thread_set_exception_ports)))
#endif ;; ENABLE(BLOCK_SET_EXCEPTION_PORTS)
#endif ;; HAVE(HARDENED_MACH_EXCEPTIONS)

kmiller68 reviewed (approved):
r=me on C++ side

Commit:
new exception handler adoption
https://bugs.webkit.org/show_bug.cgi?id=269728
rdar://119256062

Reviewed by Keith Miller.

This adopts the new hardened mach exception mechanism which was
introduced in darwin recently. See the radar for more details.

* Source/JavaScriptCore/llint/LLIntData.cpp:
* Source/JavaScriptCore/llint/LLIntData.h:
* Source/JavaScriptCore/runtime/MachineContext.h:
(JSC::MachineContext::setInstructionPointer):
* Source/JavaScriptCore/wasm/WasmFaultSignalHandler.cpp:
(JSC::Wasm::trapHandler):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/PlatformJSCOnly.cmake:
* Source/WTF/wtf/PlatformMac.cmake:
* Source/WTF/wtf/PlatformRegisters.h:
* Source/WTF/wtf/threads/Signals.cpp:
(WTF::initMachExceptionHandlerThread):
(WTF::setExceptionPorts):
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:

Canonical link: https://commits.webkit.org/276198@main
